test/src/test/java/org/apache/accumulo/test/BatchWriterInTabletServerIT.java

```diff
@@ -0,0 +1,120 @@
+    String[] uniqueNames = getUniqueNames(2);
+    String t1 = uniqueNames[0], t2 = uniqueNames[1];
+    Connector c = getConnector();
+    int numEntriesToWritePerEntry = 50;
+    IteratorSetting itset = BatchWriterIterator.iteratorSetting(6, 0, 15, 1000, numEntriesToWritePerEntry, t2, c, getToken(), false, false);
+    test(t1, t2, c, itset, numEntriesToWritePerEntry);
+  }
+
+  /**
+   * Fixed by ACCUMULO-4229.
+   * <p>
+   * This tests a situation that a client which shares a LocatorCache with the tablet server may fall into.
+   * Before the problem was fixed, adding a split after the Locator cache falls out of sync caused the
+   * BatchWriter to continuously attempt to write to an old, closed tablet.
+   * It would do so for 15 seconds until a timeout on the BatchWriter.
+   */
+  @Test
+  public void testClearLocatorAndSplitWrite() throws Exception {
+    String[] uniqueNames = getUniqueNames(2);
+    String t1 = uniqueNames[0], t2 = uniqueNames[1];
+    Connector c = getConnector();
+    int numEntriesToWritePerEntry = 50;
+    IteratorSetting itset = BatchWriterIterator.iteratorSetting(6, 0, 15, 1000, numEntriesToWritePerEntry, t2, c, getToken(), true, true);
+    test(t1, t2, c, itset, numEntriesToWritePerEntry);
+  }
+
+  private void test(String t1, String t2, Connector c, IteratorSetting itset, int numEntriesToWritePerEntry) throws Exception {
+    // Write an entry to t1
+    c.tableOperations().create(t1);
+    Key k = new Key(new Text("row"), new Text("cf"), new Text("cq"));
+    Value v = new Value("1".getBytes());
+    {
+      BatchWriterConfig config = new BatchWriterConfig();
+      config.setMaxMemory(0);
+      BatchWriter writer = c.createBatchWriter(t1, config);
+      Mutation m = new Mutation(k.getRow());
+      m.put(k.getColumnFamily(), k.getColumnQualifier(), v);
+      writer.addMutation(m);
+      writer.close();
+    }
+
+    // Create t2 with a combiner to count entries written to it
+    c.tableOperations().create(t2);
+    IteratorSetting summer = new IteratorSetting(2, "summer", SummingCombiner.class);
+    LongCombiner.setEncodingType(summer, LongCombiner.Type.STRING);
+    LongCombiner.setCombineAllColumns(summer, true);
+    c.tableOperations().attachIterator(t2, summer);
+
+    Map.Entry<Key,Value> actual;
+    // Scan t1 with an iterator that writes to table t2
+    Scanner scanner = c.createScanner(t1, Authorizations.EMPTY);
+    scanner.addScanIterator(itset);
+    actual = Iterators.getOnlyElement(scanner.iterator());
+    Assert.assertTrue(actual.getKey().equals(k, PartialKey.ROW_COLFAM_COLQUAL));
+    Assert.assertEquals(BatchWriterIterator.SUCCESS_VALUE, actual.getValue());
+    scanner.close();
+
+    // ensure entries correctly wrote to table t2
+    scanner = c.createScanner(t2, Authorizations.EMPTY);
+    actual = Iterators.getOnlyElement(scanner.iterator());
+    // System.out.println("t2 entry is " + actual.getKey().toStringNoTime() + " -> " + actual.getValue());
```

Comment:
If this is going to help debug the test on someone else's system, uncomment it and make it a logger debug. Otherwise, remove it. Commented debug statements are useless.